adapters/krushmedia/krushmedia.go

```diff
@@ -0,0 +1,197 @@
+package krushmedia
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"text/template"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type KrushmediaAdapter struct {
+	endpoint template.Template
+}
+
+func NewKrushmediaBidder(endpointTemplate string) *KrushmediaAdapter {
+	template, err := template.New("endpointTemplate").Parse(endpointTemplate)
+	if err != nil {
+		return nil
+	}
+	return &KrushmediaAdapter{endpoint: *template}
+}
+
+func checkHasImps(request *openrtb.BidRequest) error {
+	if len(request.Imp) == 0 {
+		err := &errortypes.BadInput{
+			Message: "Missing Imp Object",
+		}
+		return err
+	}
+	return nil
+}
+
+func getHeaders(request *openrtb.BidRequest) *http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("X-Openrtb-Version", "2.5")
+
+	if request.Device != nil {
+		if len(request.Device.UA) > 0 {
+			headers.Add("User-Agent", request.Device.UA)
+		}
+
+		if len(request.Device.IP) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+		}
+
+		if len(request.Device.Language) > 0 {
+			headers.Add("Accept-Language", request.Device.Language)
+		}
+
+		if request.Device.DNT != nil {
+			headers.Add("Dnt", strconv.Itoa(int(*request.Device.DNT)))
+		}
+	}
+
+	return &headers
+}
+
+func (a *KrushmediaAdapter) MakeRequests(
+	openRTBRequest *openrtb.BidRequest,
+	reqInfo *adapters.ExtraRequestInfo,
+) (
+	requestsToBidder []*adapters.RequestData,
+	errs []error,
+) {
+
+	request := *openRTBRequest
+
+	if err := checkHasImps(&request); err != nil {
+		return nil, []error{err}
+	}
+
+	var errors []error
+	var krushmediaExt *openrtb_ext.ExtKrushmedia
+	var err error
+
+	for i, imp := range request.Imp {
+		krushmediaExt, err = a.getImpressionExt(&imp)
+		if err != nil {
+			errors = append(errors, err)
+			break
+		}
+		request.Imp[i].Ext = nil
+	}
```

Comment:
Under this logic we'll keep and pass to `url, err := a.buildEndpointURL(krushmediaExt)` in line 98 the extension of the last element of the `request.Imp` array. Did we intend this?  What if the Ext's `AccountID` differs from one `Imp.Ext` element to another? 
```
request.Imp { {ID:"imp1",Ext:'{"key":"val"}'}, {ID:"imp2",Ext:'{"key":"differentValue"}'} }
```
If that is not the case and every single `Imp.Ext` element comes with the exact same `AccountID` as ilustrated in the example below:
```
request.Imp { {ID:"imp1",Ext:'{"key":"sameVal"}'}, {ID:"imp2",Ext:'{"key":"sameVal"}'} }
```
Can't we optimize by not calling `a.getImpressionExt(&imp)` on every single one of them? (And rewriting `krushmediaExt` everytime) We could just parse the first element:
```
 66   func (a *KrushmediaAdapter) MakeRequests(
 67       openRTBRequest *openrtb.BidRequest,
 68       reqInfo *adapters.ExtraRequestInfo,
 69   ) (
 70       requestsToBidder []*adapters.RequestData,
 71       errs []error,
 72   ) {
 73   *-- 10 lines: request := *openRTBRequest--------------------
 83  
    +     krushmediaExt, err = a.getImpressionExt(request.Imp[0])
    +     if err != nil {
    +         return nil, append(errors, err)
    +     }
    +
 84       for i, imp := range request.Imp {
 85 -         krushmediaExt, err = a.getImpressionExt(&imp)
 86 -         if err != nil {
 87 -             errors = append(errors, err)
 88 -             break
 89 -         }
 90           request.Imp[i].Ext = nil
 91       }
 92  
 93 -     if len(errors) > 0 {
 94 -         return nil, errors
 95 -     }
 96  
 97       url, err := a.buildEndpointURL(krushmediaExt)
 98       if err != nil {
 99           return nil, []error{err}
100       }
101   *--  6 lines: reqJSON, err := json.Marshal(request)---------
107       return []*adapters.RequestData{{
108           Method:  http.MethodPost,
109           Body:    reqJSON,
110           Uri:     url,
111           Headers: *getHeaders(&request),
112       }}, nil
113   }
adapters/krushmedia/krushmedia.go [+]
```